rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,110 @@
+---
+authors: Noah Stride <noah@goteleport.com>
+state: draft
+---
+
+# RFD 79 - OIDC JWT Joining
+
+## Required Approvers
+
+* Engineering @zmb3 && @nklaassen
+* Security @reedloden
+* Product: (@xinding33 || @klizhentas)
+
+## Terminology
+
+- OIDC: OpenID Connect. A federated authentication protocol built on top of OAuth 2.0.
+- OP: OpenID Provider.
+- Issuer: an OP that has issued a specific token.
+- Workload Identity: an identity corresponding to a running workload, such as a container or VM. This is in contrast to the more well-known use of OAuth in identifying users. A workload identity token is often made available to a workload running on a platform via a metadata server or an environment variable.
+
+## What
+
+Teleport should support trusting a third party OP and the JWTs that it issues when authenticating a client for the cluster join process. This is similar to the support for IAM joining, and will allow joining a Teleport cluster without the need to distribute a token on several platforms.
+
+Users will need to be able to configure trust in an OP, and rules that determine what identities are allowed to join the cluster.
+
+## Why
+
+This feature reduces the friction involved in adding many new nodes to Teleport on a variety of platforms. This is also more secure, as the user does not need to distribute a token which is liable to exfilitration.
+
+Whilst multiple providers offer OIDC identities to workloads running on their platform, we will start by targetting GCP GCE since this represents a large portion of the market. However, the work towards this feature will also enable us to simply add other providers that support OIDC such as:
+
+- GitHub Actions: a key platform for growing usage of Machine ID.
+- GitLab CI/CD
+- GCP GCB
+
+## Details
+
+The work on OIDC joining is broken down into two parts:
+
+- Support in the Auth server for trusting an OP issued JWT. This work is general, and will be applicable to all providers.
+- Support in nodes for fetching their workload identity token from their environment. This work will be specific to each platform we intend to support.
+
+OIDC supports multiple types of token (`id_token`: a JWT encoding information about the identity, which can be verified using the issuer's public key and `access_token`: an opaque token that can be used with a `userinfo` endpoint on the issuer to obtain information about the identity). However, in the case of workload identities, `id_token` is the most prevelant. For this reason, our initial implementation will solely support `id_token`.
+
+### Auth server support
+
+#### Configuration
```

Comment:
One of my largest concerns with this plan, is that this generic configuration with CEL is too flexible.

This causes me a concern for two reasons:
- Complexity. Users will need to find documentation that explains what claims are available for the provider they intend to trust. We may be able to provide this documentation for a few key providers, but we risk falling behind changes.
- Security. It is much harder for us to guide users towards creating secure rules. If we made our configuration more provider-specific, we'd be able to easily ensure they were creating rules that included at least one claim that was scoped to their project/organisation on that platform.

On the other hand this flexibility allows users to come up with much more complicated rules, which may be appropriate for complicated cloud projects, and the generic configuration makes it much easier for us to add support for additional providers in future.

The alternative would be to have something along the lines of:

```yaml
kind: token
version: v2
metadata:
  name: my-gce-token
  expires: "3000-01-01T00:00:00Z"
spec:
  roles: [Node]
  join_method: gce
  allow: 
  - project_id: "my-project"
     instance_name: "an-instance"
```

Which is more similar to how we support AWS.

We'd be able to share much of the implementation code between providers that use OIDC, but would still end up needing to build in auth server support for each provider.

Reply:
Sure, although, what I had in mind was that if we were going to make this more specific to providers, we would want to do some mapping/extraction. For example, the useful fields in the GCP claims are buried quite deep:

```json
{
  "aud": "https://noah.teleport.sh",
  "azp": "REDACTED",
  "email": "REDACTED-compute@developer.gserviceaccount.com",
  "email_verified": true,
  "exp": 1657540679,
  "google": {
    "compute_engine": {
      "instance_creation_timestamp": 1657536818,
      "instance_id": "REDACTED",
      "instance_name": "test-instance-noah",
      "project_id": "REDACTED",
      "project_number": REDACTED,
      "zone": "us-central1-a"
    }
  },
  "iat": 1657537079,
  "iss": "https://accounts.google.com",
  "sub": "REDACTED"
}
```

If we were doing a provider specific implementation, we could pull `google.compute_engine.project_id` up to the top level of the matcher, and even add additional validation rules to ensure that the user has provided a filter on at least one field that makes it secure (e.g subject, email, project number/id).